R/fmelt.R

```diff
@@ -3,19 +3,19 @@
 #   reshape2 package is deprecated since December 2017, so we'll deprecate our
 #   redirection as well
 
-melt <- function(data, ..., na.rm = FALSE, value.name = "value") {
-  if (is.data.table(data)) {
-    UseMethod("melt", data)
-    # if data is not data.table and reshape2 is installed, this won't dispatch to reshape2's method;
-    # CRAN package edarf and others fail without the else branch
+melt = function(data, ..., na.rm = FALSE, value.name = "value") {
+  UseMethod("melt", data)
+}
+
+melt.default = function(data, ..., na.rm = FALSE, value.name = "value") {
+  # if no registered method exists for data, attempts to redirect data to reshape2::melt;
+  # CRAN package edarf and others fail without the redirection
   # nocov start
-  } else {
-    data_name = deparse(substitute(data))
-    ns = tryCatch(getNamespace("reshape2"), error=function(e)
-      stop("The melt generic in data.table has been passed a ",class(data)[1L],", but data.table::melt currently only has a method for data.tables. Please confirm your input is a data.table, with setDT(", data_name, ") or as.data.table(", data_name, "). If you intend to use a method from reshape2, try installing that package first, but do note that reshape2 is deprecated and you should be migrating your code away from using it."))
-    warning("The melt generic in data.table has been passed a ", class(data)[1L], " and will attempt to redirect to the relevant reshape2 method; please note that reshape2 is deprecated, and this redirection is now deprecated as well. To continue using melt methods from reshape2 while both libraries are attached, e.g. melt.list, you can prepend the namespace like reshape2::melt(", data_name, "). In the next version, this warning will become an error.")
-    ns$melt(data, ..., na.rm=na.rm, value.name=value.name)
-  }
+  data_name = deparse(substitute(data))
+  ns = tryCatch(getNamespace("reshape2"), error=function(e)
+    stop("The melt generic in data.table has been passed a ", class(data)[1L],", but data.table::melt currently only has a method for data.tables. Please confirm your input is a data.table, with setDT(", data_name, ") or as.data.table(", data_name, "). If you intend to use a method from reshape2, try installing that package first, but do note that reshape2 is deprecated and you should be migrating your code away from using it."))
+  warning("The melt generic in data.table has been passed a ", class(data)[1L], " and will attempt to redirect to the relevant reshape2 method; please note that reshape2 is deprecated, and this redirection is now deprecated as well. To continue using melt methods from reshape2 while both libraries are attached, e.g. melt.list, you can prepend the namespace like reshape2::melt(", data_name, "). In the next version, this warning will become an error.")
```

Comment:
I think we are no going into new major version so this code could be simplified for raising error, as mentioned in the warning already.

Reply:
Do you mean stopping the redirection towards `reshape2`?